### PR TITLE
nanoAOD: fix electron IP sign

### LINKED ISSUE #514: Electron IP sign
In #479 the sign of the 2D IP (dxy) for muons was fixed, and if I understand correctly the issue was fixed upstream in PAT (here https://github.com/cms-sw/cmssw/pull/28753) by switching from using `IPTools::signedTransverseImpactParameter()` to ` reco::TrackBase::dxy()`.

However it seems that PAT electrons still use `signedTransverseImpactParameter()`: https://github.com/cms-nanoAOD/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/PATElectronProducer.cc#L1150-L1159

Is there a reason why this wasn't changed for electrons?

### PR DESCRIPTION
**Pull request by swertz**
#### PR description:

This is the correspondence of https://github.com/cms-nanoAOD/cmssw/pull/479 but for electrons.

The sign in miniAOD was fixed in #28753 for muons and #30738 for electrons. This PR fixes the sign at nanoAOD level when running on existing miniAOD, where the fix is not present yet.

Will need a backport to 106X.

Closes https://github.com/cms-nanoAOD/cmssw/issues/514

#### PR validation:

Validated by processing 2000 ttbar events on RunIISummer16MiniAODv3 (with `auto:run2_mc` and `Run2_2016,run2_nanoAOD_94X2016`) and RunIIAutumn18MiniAOD (with `auto:phase1_2018_realistic` and `Run2_2018,run2_nanoAOD_102Xv1`): the only electron variable that shows any change is `dxy`, and as expected only the sign changes:

Electron_dxy_18

### CODE REVIEW

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
@swertz 

What about the miniAOD-UL of type `'run2_nanoAOD_106Xv1'`
The IP will get fixed for muon and electron only for the newly produced miniAOD (from master here or those with the run2_miniAOD_UL in 10_6)

**Comment by swertz** [comment]
> What about the miniAOD-UL of type `'run2_nanoAOD_106Xv1'`
> The IP will get fixed for muon and electron only for the newly produced miniAOD (from master here or those with the run2_miniAOD_UL in 10_6)

Indeed... but then it seems the `run2_nanoAOD_106Xv1` modifier needs to be added everywhere for electrons and muons (and elsewhere?), which goes beyond the scope of the electron IP fix in this PR. Perhaps this should be taken care of in another PR (by the electron or muon POG, as there might be other changes needed for 106X that I would overlook)?

**Comment by santocch** [comment]
+1

**Comment by silviodonato** [comment]
@mariadalfonso are there other modifiers to be considered beyond `run2_nanoAOD_106Xv1`?

If it is just matter of adding `run2_nanoAOD_106Xv1` to the list `run2_nanoAOD_94XMiniAODv1,run2_nanoAOD_94XMiniAODv2,run2_nanoAOD_94X2016,run2_nanoAOD_102Xv1` I would encourage @swertz  to add this fix in this PR. 

We can leave the muon fix in a separated PR.

**Comment by mariadalfonso** [comment]
indeed adding this run2_nanoAOD_106Xv1 should be enough.
@afiqaize from gamma promised to have a look and confirm the changes and get back to @swertz 

> We can leave the muon fix in a separated PR.
agreed

**Comment by afiqaize** [comment]
fixing the sign at nano level is possible with minor modifications to PhysicsTools/PatAlgos/plugins/LeptonUpdater.cc, with the fixDxySign flag set to true. modifications is needed because bestTrack() is not the same as gsfTrack().

I checked that this gives the same result as running on mini with the fix applied, except some precision mismatch in 2% of the cases (tested 6k bb4l events). I did not manage to find out why this is happening, but the effect is O(0.001) which I guess is fine.

how should this be propagated? given that my copy PhysicsTools/NanoAOD/python/electrons_cff.py is changed such that it runs only for this particular check, I prefer that @swertz takes over [1] and handles the proper nano-specific era modifications etc.

[1] /afs/cern.ch/work/a/afiqaize/public/codes/misc/LeptonUpdater.cc

**Comment by swertz** [comment]
> modifications is needed because bestTrack() is not the same as gsfTrack()

Thanks, I had overlooked that. I've pulled the suggested changes to `LeptonUpdater.cc`. 

@afiqaize  About the era modifications: is it OK for you to add `run2_nanoAOD_106Xv1` to _all_ modifier lists where `run2_nanoAOD_102Xv1` was already present or are there exceptions to that?

**Comment by afiqaize** [comment]
@swertz I think it's fine, to my knowledge only the magnitude is used in selections, and if one needs the signs the fixed ones are more correct.

**Comment by swertz** [comment]
What I meant is, can `run2_nanoAOD_106Xv1` be added everywhere in `electrons_cff.py` where `run2_nanoAOD_102Xv1` is already present, i.e. also for things not related to the IP sign?

**Comment by afiqaize** [comment]
I'm afraid I don't understand the question. Are you asking if it's ok to fix the sign everywhere, on top of other modifiers? If so, I'd say yes.

Otherwise I'll need some further clarifications.

**Comment by swertz** [comment]
At this point the electron modifiers in nanoAOD (`electrons_cff.py`) have not been updated with `run2_nanoAOD_106Xv1`, see e.g. in https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/python/electrons_cff.py#L132 but everywhere else as well.

I can add the `run2_nanoAOD_106Xv1` modifier for what concerns the IP sign fix, but it was suggested above that we take the opportunity of this PR to add the `run2_nanoAOD_106Xv1` modifier everywhere else. But I'd like to make sure this has no unexpected consequence...

**Comment by afiqaize** [comment]
Ok, now I see. I'm afraid I don't know; initially I tried to do it that way but due to time constraints I just broke the cff such that the leptonupdater always runs, and every other thing not in the core patelectron class doesn't.

So I can only recommend trying it and see what happens. It may even be useful, unexpected mismatch means unexpected usage of the sign information, which is an interesting information.

**Comment by swertz** [comment]
I've added the `run2_nanoAOD_106Xv1` modifier everywhere for **electrons** and I've checked again that all the variables come out the same (except for the IP sign).

File: PhysicsTools/PatAlgos/plugins/LeptonUpdater.cc [hunk 1]
```diff
@@ -181,9 +181,15 @@ void pat::LeptonUpdater<T>::produce(edm::StreamID, edm::Event &iEvent, edm::Even
       float signPV = 1.;
       float signBS = 1.;
       if (beamSpotIsValid) {
-        signBS = copysign(1., lep.bestTrack()->dxy(beamSpot));
+        if constexpr (std::is_same_v<T, pat::Muon>)
```
Comment (perrotta) [inline PhysicsTools/PatAlgos/plugins/LeptonUpdater.cc:184]: I would rather consider the `pat::electron` as the "special case", to be considered separately from "all other cases" (ok, by now "all other cases" is just `pat::muon` indeed).
Possible suggestion:
```C++
        if constexpr (std::is_same_v<T, pat::Electron>)
          signBS = copysign(1., lep.gsfTrack()->dxy(beamSpot));
        else
          signBS = copysign(1., lep.bestTrack()->dxy(beamSpot));
```
(and the same for `signPV` here below)

Comment (swertz) [reply PhysicsTools/PatAlgos/plugins/LeptonUpdater.cc]: Thanks, good point

**Comment by perrotta** [comment]
please test

**Comment by mariadalfonso** [comment]
difference are spotted in dxy as expected

all_OldVSNew_TTbar13nanoEDMin2017wf1325p7c_nanoaodFlatTable_electronTable__DQM_obj_floats__dxy_18

**Comment by mariadalfonso** [comment]
please test workflow 1325.61, 1325.81

(those are TTbar related to run2_nanoAOD_106Xv1)

**Comment by mariadalfonso** [comment]
+xpog

dxy updated in nano for the old miniAOD, this is be properly fixed in miniAOD UL

**Comment by perrotta** [comment]
@swertz , in spite of the changes in PhysicsTools/PatAlgos/plugins/LeptonUpdater.cc I don't see any difference in the miniAOD outputs for the automatic tests. It is true that the possible updates are only for the signs, therefore they could only show up with large stat: but did you ever check that some difference show up from time to time?
(Just wondering: isn't the `bestTrack()` the same as the `gsfTrack()` for the electrons?)

**Comment by swertz** [comment]
> in spite of the changes in PhysicsTools/PatAlgos/plugins/LeptonUpdater.cc I don't see any difference in the miniAOD outputs for the automatic tests. It is true that the possible updates are only for the signs, therefore they could only show up with large stat: but did you ever check that some difference show up from time to time?

But the changes should only show up in nanoAOD, see the plot @mariadalfonso posted above. This is a nanoAOD-level fix for old miniAODs where the bug is present. New miniAODs will have the sign fixed from #30738.

> (Just wondering: isn't the bestTrack() the same as the gsfTrack() for the electrons?)

OK, this time I actually checked the code... From [this](http://cmsdoxygen.web.cern.ch/cmsdoxygen/CMSSW_11_2_0_pre7/doc/html/d2/df5/classreco_1_1RecoCandidate.html#aa13a56c5ec2e398d5d430d3f48e9ed33) and [this](http://cmsdoxygen.web.cern.ch/cmsdoxygen/CMSSW_11_2_0_pre7/doc/html/d2/d1f/classpat_1_1Electron.html#a26b625bf9312ba10e3a0cc65398494d2) it seems you are right, @afiqaize can you confirm? If that's the case the extra condition changes nothing, would you prefer removing it?

**Comment by afiqaize** [comment]
I didn't check the code; my validation has been at event level: unfixed mini -> nano gave different dxy signs wrt fixed mini. and applying the changes makes them agree.

in both cases I made the mini from aod without and with the UL modifier, within the same setup. from the code you link it seems that they're the same, but I can't comment on what happens within the chain that led to my observations. which if I understand correctly is replicated by the nanoDQM check.

**Comment by mariadalfonso** [comment]
-xpog

another things that I noticed is that in the run2_nanoAOD_106Xv1 also the scale of the electron change.
We need e-gamma to approve these line changes
https://github.com/cms-sw/cmssw/pull/31495/files#diff-170f93a54063f185360137b9496b90caR503-R514

Screen Shot 2020-10-06 at 22 19 56

**Comment by afiqaize** [comment]
do I understand correctly that the modifier runs more things than what's needed for the dxy sign fix? if so, I think the changes go beyond the scope of the PR.

**Comment by mariadalfonso** [comment]
> do I understand correctly that the modifier runs more things than what's needed for the dxy sign fix? if so, I think the changes go beyond the scope of the PR.

yes but let's understand first what to do with the corrections, then we decide if we have in a follow up PR or this one.
There are corrections applied to all the nano except to those from run2_nanoAOD_106Xv1: where they forgotten or should not be applied ?

File: PhysicsTools/NanoAOD/python/electrons_cff.py [hunk 2]
```diff
@@ -270,11 +263,12 @@ def _get_bitmapVIDForEle_docstring(modules,WorkingPoints):
...
 )
 
-run2_nanoAOD_102Xv1.toModify(slimmedElectronsWithUserData.userFloats,
-    ecalTrkEnergyErrPostCorrNew = cms.InputTag("calibratedPatElectrons102X","ecalTrkEnergyErrPostCorr"),
-    ecalTrkEnergyPreCorrNew     = cms.InputTag("calibratedPatElectrons102X","ecalTrkEnergyPreCorr"),
-    ecalTrkEnergyPostCorrNew    = cms.InputTag("calibratedPatElectrons102X","ecalTrkEnergyPostCorr"),
-)
+for modifier in run2_nanoAOD_102Xv1,run2_nanoAOD_106Xv1:
+    modifier.toModify(slimmedElectronsWithUserData.userFloats,
+        ecalTrkEnergyErrPostCorrNew = cms.InputTag("calibratedPatElectrons102X","ecalTrkEnergyErrPostCorr"),
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/electrons_cff.py:268]: @swertz 
Should this be calibratedPatElectrons102X for 102X and  calibratedPatElectrons106X for 106X ?

**Comment by swertz** [comment]
> another things that I noticed is that in the run2_nanoAOD_106Xv1 also the scale of the electron change.

That's puzzling, the comparisons I ran locally with `run2_nanoAOD_106Xv1` (over `RunIIAutumn18MiniAOD`) didn't show any changes in electron scale:

Electron_pt
Electron_scEtOverPt

Note that in my comparison, the baseline was produced using `run2_nanoAOD_102Xv1`. Could the reason be that in your DQM link, the baseline is produced with the `run2_nanoAOD_106Xv1` modifier? Since without this PR this modifier is entirely missing for the electrons, it means it would not use the expected corrections but the "default" ones (see [this](https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/python/electrons_cff.py#L185-L189) and [this](https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaTools/python/calibratedEgammas_cff.py#L9)).

> do I understand correctly that the modifier runs more things than what's needed for the dxy sign fix? if so, I think the changes go beyond the scope of the PR.

Well, this is what I meant when I wrote earlier: https://github.com/cms-sw/cmssw/pull/31495#issuecomment-703467101

**Comment by mariadalfonso** [comment]
@swertz 

the right comparison should be using the inputs as defined by the 'run2_nanoAOD_106Xv1' w/ and w/o this PR.
run2_nanoAOD_106Xv1 points to miniAOD inputs that have been made out of Summer19 campaign that are already with UL reco, but prior to the reminiAOD that is planned.

Indeed the reason why we have this difference is that the correction is missing completely for the run2_nanoAOD_106Xv1.
I would like to know by the egamma if this was intended or an oversight and need a final sign of from them on this aspect of the PR.

**Comment by perrotta** [comment]
> But the changes should only show up in nanoAOD, see the plot @mariadalfonso posted above. This is a nanoAOD-level fix for old miniAODs where the bug is present. New miniAODs will have the sign fixed from #30738.
> 

Ok, I overlooked the fact that the `LeptonUpdater` only enters in nanoAOD production, and not in miniAOD
Still, please evaluate whether the changes implemented for LeptonUpdater.cc are needed or suitable

**Comment by swertz** [comment]
> the right comparison should be using the inputs as defined by the 'run2_nanoAOD_106Xv1' w/ and w/o this PR.

Yes, that was my mistake. I have now attempted to remove anything not related to the IP sign fix. Egamma can then take care of adding the `run2_nanoAOD_106Xv1` modifier where needed otherwise (e.g. for scale corrections). However with the current version I keep getting an error I don't understand:
```
  File "/work/swertz/CMSSW_11_2_0_pre6_NanoAOD/python/PhysicsTools/NanoAOD/nano_cff.py", line 222, in nanoAOD_activateVID
    process.electronSequence.insert(process.electronSequence.index(process.bitmapVIDForEle),process.egmGsfElectronIDSequence)
  File "/cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre6/python/FWCore/ParameterSet/SequenceTypes.py", line 469, in index
    return self._seq.index(item)
  File "/cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre6/python/FWCore/ParameterSet/SequenceTypes.py", line 193, in index
    return self._collection.index(item)
ValueError: cms.EDProducer("EleVIDNestedWPBitmapProducer",
    WorkingPoints = cms.vstring(
        'egmGsfElectronIDs:cutBasedElectronID-Fall17-94X-V2-veto', 
        'egmGsfElectronIDs:cutBasedElectronID-Fall17-94X-V2-loose', 
        'egmGsfElectronIDs:cutBasedElectronID-Fall17-94X-V2-medium', 
        'egmGsfElectronIDs:cutBasedElectronID-Fall17-94X-V2-tight'
    ),
    src = cms.InputTag("slimmedElectronsUpdated")
)
 is not in list
```
even though `bitmapVIDForEle` is in the process...

>  Still, please evaluate whether the changes implemented for LeptonUpdater.cc are needed or suitable

I have checked again with and without the use of `gsfTrack()` and indeed see strictly no difference. Not sure what @afiqaize did different...

**Comment by afiqaize** [comment]
> I have checked again with and without the use of `gsfTrack()` and indeed see strictly no difference. Not sure what @afiqaize did different...

odd. maybe there was a mistake in how I put the cff together. anyway I won't be able to look at it anytime soon, so if you're sure there's no difference I'm ok with removing the condition.

personally however, I find the code clearer with it.

**Comment by perrotta** [comment]
> > I have checked again with and without the use of `gsfTrack()` and indeed see strictly no difference. Not sure what @afiqaize did different...
> 
> odd. maybe there was a mistake in how I put the cff together. anyway I won't be able to look at it anytime soon, so if you're sure there's no difference I'm ok with removing the condition.
> 
> personally however, I find the code clearer with it.

Fine with me: if the message is "whatever will be chosen in the future for the `bestTrack()`, for the electrons I want to stick on the sign as given by the `gsfTrack()`" better to let it as you have coded it

**Comment by perrotta** [comment]
@swertz please notify here when you think that the "error I don't understand" that you mention in https://github.com/cms-sw/cmssw/pull/31495#issuecomment-704850586 can be considered as fixed, so that we can launch the tests

**Comment by swertz** [comment]
With the latest version I believe it works.

I had to replace
```
_withUpdate_sequence = cms.Sequence(slimmedElectronsUpdated + electronSequence.copy())
```
by
```
_withUpdate_sequence = electronSequence.copy()
_withUpdate_sequence.replace(bitmapVIDForEle, slimmedElectronsUpdated + bitmapVIDForEle)
```
But I'm not sure why the former would not work...

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
> 
> But I'm not sure why the former would not work...

@swertz did you try to dump the explicit content of the sequences in the two cases?

**Comment by mariadalfonso** [comment]
I confirms that the electron now shows changes only on the day as wanted in all previous eras

**Comment by mariadalfonso** [comment]
+xpog

electron IP fixed on the fly on previously made miniAOD
need to follow up on gamma on the situation of the scale correction for the 10_6 mini based on run2_nanoAOD_106Xv1

**Comment by perrotta** [comment]
+1
- Everything coded in this PR ends up in the nanoAOD
- For reco the code is fine (although I would have personally named in a less complicate way the new sequences defined or replaced in the python configs...)
- Jenkis tests pass and only show differences in nanoAOD, as expected

**Comment by mariadalfonso** [comment]
@swertz please make backport of this PR

**Comment by silviodonato** [comment]
merge

**Comment by santocch** [comment]
+1

**Pull request merged**